src/ethereum/berlin/vm/instructions/log.py

```diff
@@ -50,37 +43,21 @@ def log_n(evm: Evm, num_topics: U256) -> None:
     :py:class:`~ethereum.berlin.vm.exceptions.StackUnderflowError`
         If `len(stack)` is less than `2 + num_topics`.
     """
-    ensure(not evm.message.is_static, WriteInStaticContext)
-    # Converting memory_start_index to Uint as memory_start_index + size - 1
-    # can overflow U256.
-    memory_start_index = Uint(pop(evm.stack))
+    # STACK
+    memory_start_index = pop(evm.stack)
     size = pop(evm.stack)
 
-    gas_cost_log_data = u256_safe_multiply(
-        GAS_LOG_DATA, size, exception_type=OutOfGasError
-    )
-    gas_cost_log_topic = u256_safe_multiply(
-        GAS_LOG_TOPIC, num_topics, exception_type=OutOfGasError
-    )
-    gas_cost_memory_extend = calculate_gas_extend_memory(
-        evm.memory, memory_start_index, size
-    )
-    gas_cost = u256_safe_add(
-        GAS_LOG,
-        gas_cost_log_data,
-        gas_cost_log_topic,
-        gas_cost_memory_extend,
-        exception_type=OutOfGasError,
-    )
-    evm.gas_left = subtract_gas(evm.gas_left, gas_cost)
-
-    extend_memory(evm.memory, memory_start_index, size)
-
     topics = []
     for _ in range(num_topics):
         topic = pop(evm.stack).to_be_bytes32()
         topics.append(topic)
 
+    # GAS
+    extend_memory(evm, memory_start_index, size)
+    charge_gas(evm, GAS_LOG + GAS_LOG_DATA * size + GAS_LOG_TOPIC * num_topics)
+
+    # OPERATION
+    ensure(not evm.message.is_static, WriteInStaticContext)
```

Comment:
Changing the order of this is fine, because there's no observable difference between an `OutOfGasError` and a `WriteInStaticContext`?

Reply:
Correct. We never distinguish between subclasses of `ExceptionalHalt`.